### ISSUE TITLE
How to get the AP with ILSVRC2013 datasets?

### ISSUE THREAD

**Issue by mickey0217** [rbgirshick/py-faster-rcnn#334]
I finally trained the ILSVRC2013 DET datasets(200 classess) with VGG16 and got the trained models.
Strange as it may sound,
I want to get AP of each class on the terminal or some other method that can check directly:) , but I failed to do that. 

Is there any body who already do that successfully with the same datasets as I use?
I use the ILSVRC 2013 Validation sets to get mAP result. 

**Because I'm a beginner who have just started faster rcnn and i'm not a good programmer. 
Maybe I could do commit an elementary mistake. **so, please let's me know any good idea even if you don't use the datasets same as me for testing.**** thank you:)!
   
____my partial code (I think I have difficulty in comparing detection results and gt results. )_____

[in test.py]_____________________________________________

def test_net(net, imdb, max_per_image=100, thresh=0.05, vis=False): 
    """Test a Fast R-CNN network on an image database."""
    num_images = len(imdb.image_index)

    all_boxes = [[[] for _ in xrange(num_images)]
                 for _ in xrange(imdb.num_classes)]

    output_dir = get_output_dir(imdb, net)

    # timers
    _t = {'im_detect' : Timer(), 'misc' : Timer()}

    if not cfg.TEST.HAS_RPN:
        roidb = imdb.roidb

    for i in xrange(num_images):
        # filter out any ground truth boxes
        if cfg.TEST.HAS_RPN:
            box_proposals = None
        else:
            # The roidb may contain ground-truth rois (for example, if the roidb
            # comes from the training or val split). We only want to evaluate
            # detection on the *non*-ground-truth rois. We select those the rois
            # that have the gt_classes field set to 0, which means there's no
            # ground truth.
            box_proposals = roidb[i]['boxes'][roidb[i]['gt_classes'] == 0]

        im = cv2.imread(imdb.image_path_at(i))
        _t['im_detect'].tic()
        scores, boxes = im_detect(net, im, box_proposals)
        _t['im_detect'].toc()

        _t['misc'].tic()
        # skip j = 0, because it's the background class
        for j in xrange(1, imdb.num_classes):
            inds = np.where(scores[:, j] > thresh)[0]
            cls_scores = scores[inds, j]
            cls_boxes = boxes[inds, j*4:(j+1)*4]
            cls_dets = np.hstack((cls_boxes, cls_scores[:, np.newaxis])) \
                .astype(np.float32, copy=False) # make cls_boxes and cls_scores into the one array 
            keep = nms(cls_dets, cfg.TEST.NMS)
            cls_dets = cls_dets[keep, :]
            dets = cls_dets
            classname = imdb.classes[j]
            if vis == True:
                vis_detections(im, imdb.classes[j], cls_dets)

            else : 
                predicted_class = vis_detections2(im, imdb.classes[j], cls_dets)

            all_boxes[j][i] = cls_dets

            det_file = os.path.join(output_dir,  imdb.classes[j] + '_det.pkl')

            with open(det_file, 'wb') as f:
               cPickle.dump(all_boxes, f, cPickle.HIGHEST_PROTOCOL)

   def _do_python_eval(self, output_dir = 'output'): 
        annopath = os.path.join( 
            self._devkit_path, 
            'data', 
            'Annotations', 
            '{:s}.xml') 
        imagesetfile = os.path.join( 
            self._devkit_path, 
            'data', 
            'ImageSets', 
            self._image_set + '.txt') 
        cachedir = os.path.join(self._devkit_path, 'annotations_cache') 
        print 'cachedir : %s' %cachedir 
        aps = [] 
        # The PASCAL VOC metric changed in 2010 
        #use_07_metric = True if int(self._year) < 2010 else False 
        use_07_metric = False 
        print 'VOC07 metric? ' + ('Yes' if use_07_metric else 'No') 
        if not os.path.isdir(output_dir): 
            os.mkdir(output_dir) 
        for i, cls in enumerate(self._classes): 
            if cls == '__background__': 
                continue 
            filename = self._get_ilsvrc_results_file_template(cls) 
            rec, prec, ap = voc_eval( 
                filename, annopath, imagesetfile, cls, cachedir, ovthresh=0.5, 
                use_07_metric=use_07_metric) 
            aps += [ap] 
            print('AP for {} = {:.4f}'.format(cls, ap)) 
            with open(os.path.join(output_dir, cls + '_pr.pkl'), 'w') as f: 
                cPickle.dump({'rec': rec, 'prec': prec, 'ap': ap}, f) 
        print('Mean AP = {:.4f}'.format(np.mean(aps))) 
        print('~~~~~~~~') 
        print('Results:') 
        for ap in aps: 
            print('{:.3f}'.format(ap)) 
        print('{:.3f}'.format(np.mean(aps))) 
        print('~~~~~~~~')

______________________________________________________________________

[in ilsvrc.py]

    def _do_python_eval(self, output_dir = 'output'): 
        annopath = os.path.join( 
            self._devkit_path, 
            'data', 
            'Annotations', 
            '{:s}.xml') 
        imagesetfile = os.path.join( 
            self._devkit_path, 
            'data', 
            'ImageSets', 
            self._image_set + '.txt') 
        cachedir = os.path.join(self._devkit_path, 'annotations_cache') 
        print 'cachedir : %s' %cachedir 
        aps = [] 
        # The PASCAL VOC metric changed in 2010 
        #use_07_metric = True if int(self._year) < 2010 else False 
        use_07_metric = False 
        print 'VOC07 metric? ' + ('Yes' if use_07_metric else 'No') 
        if not os.path.isdir(output_dir): 
            os.mkdir(output_dir) 
        for i, cls in enumerate(self._classes): 
            if cls == '__background__': 
                continue 
            filename = self._get_ilsvrc_results_file_template(cls) 
            rec, prec, ap = voc_eval( 
                filename, annopath, imagesetfile, cls, cachedir, ovthresh=0.5, 
                use_07_metric=use_07_metric) 
            aps += [ap] 
            print('AP for {} = {:.4f}'.format(cls, ap)) 
            with open(os.path.join(output_dir, cls + '_pr.pkl'), 'w') as f: 
                cPickle.dump({'rec': rec, 'prec': prec, 'ap': ap}, f) 
        print('Mean AP = {:.4f}'.format(np.mean(aps))) 
        print('~~~~~~~~') 
        print('Results:') 
        for ap in aps: 
            print('{:.3f}'.format(ap)) 
        print('{:.3f}'.format(np.mean(aps))) 
        print('~~~~~~~~') 
 
    def evaluate_detections(self, all_boxes, output_dir): 
        self._do_python_eval(output_dir)  
        print 'output_dir: %s' %output_dir 
        if self.config['matlab_eval']: 
            self._do_matlab_eval(output_dir) 
        if self.config['cleanup']: 
            for cls in self._classes: 
                if cls == '__background__': 
                    continue 

 
    def _get_ilsvrc_results_file_template(self,cls): 
        # VOCdevkit/results/VOC2007/Main/<comp_id>_det_test_aeroplane.txt 
        out_path = os.path.join( 
            '/home', 
            'leekaeun', 
            'py-faster-rcnn', 
            #self._devkit_path, 
            'output', 
            'faster_rcnn_end2end', 
            'val', 
            'vgg16_faster_rcnn_iter_90000/') 
        filename =  os.path.join(out_path,  cls + '_det.pkl') 
        print 'get the path of det.pkl file about each cls: %s' %cls 
        path = os.path.join(filename) 
        return path

______________________________________________________________________________

[in voc_eval.py]

def voc_eval(detpath,
             annopath,
             imagesetfile,
             classname,
             cachedir,
             ovthresh=0.5,
             use_07_metric=False):
  

    # macthing (wind <--> class name) 
    temp_mat = sio.loadmat(os.path.join(ILSVRC_devkit_path,'data','meta_det.mat'))
    win_list=[0]
    class_list=['background']
    match = 'background'
    for i in xrange(200):
        if(temp_mat['synsets'][0][i][2]) == classname:
            match = temp_mat['synsets'][0][i][1]

    # first load gt
    if not os.path.isdir(cachedir):
        os.mkdir(cachedir)
    cachefile = os.path.join(cachedir, 'annots.pkl')

    # read list of images
    with open(imagesetfile, 'r') as f: #read train.txt or test.txt 
        lines = f.readlines()
    imagenames = [x.strip() for x in lines]
    print 'succeed in getting the filename_list for test!!' 

    if not os.path.isfile(cachefile):
        # load annots
        recs = {}
 
        for i, imagename in enumerate(imagenames):
            recs[imagename] = parse_rec(annopath.format(imagename))
            # e.g. recs[imagename] :  [{'name': 'n01726692', 'bbox': [105, 204, 358, 279]}]

            if i % 100 == 0:
                print 'Reading annotation for {:d}/{:d}'.format(
                    i + 1, len(imagenames))
            print 'Reading annotation for {:d}/{:d}'.format(
            i + 1, len(imagenames))

        # save
        print 'Saving cached annotations to {:s}'.format(cachefile)
        with open(cachefile, 'w') as f:
            cPickle.dump(recs, f)
    else:
        # load
        with open(cachefile, 'r') as f:
            recs = cPickle.load(f)

    # extract gt objects for this class
    class_recs = {}
    npos = 0

    for imagename in imagenames:

        for obj in recs[imagename]:
            R = []
            print 'obj[name]: {}'.format(obj['name'])
            if obj['name'] == match:
                R.append(obj)

        bbox = np.array([x['bbox'] for x in R])
        #difficult = np.array([x['difficult'] for x in R]).astype(np.bool) #for VOC
        difficult = 0
        det = [False] * len(R)
        #npos = npos + sum(~difficult) #for VOC
        class_recs[imagename] = {'bbox': bbox,
                                 'difficult': difficult, # for voc
                                 'det': det}

        print 'class_recs[{}] : {}'.format(imagename,class_recs[imagename])

    # read dets

    detfile = detpath.format(classname)t

    with open(detfile, 'r') as f:
        lines = f.readlines()

    splitlines = [x.strip().split(' ') for x in lines]
    image_ids = [x[0] for x in splitlines]

    #confidence = np.array([float(x[1]) for x in splitlines]) 
    #BB = np.array([[float(z) for z in x[2:]] for x in splitlines]) 

    # sort by confidence -->substract!
    #sorted_ind = np.argsort(-confidence)
    #sorted_scores = np.sort(-confidence)
    #BB = BB[sorted_ind, :]
    image_ids = [image_ids[x] for x in sorted_ind]

    # go down dets and mark TPs and FPs
    nd = len(image_ids)
    tp = np.zeros(nd)
    fp = np.zeros(nd)
    for d in range(nd):
        #print 'image_ids: %s' %image_ids
        R = class_recs[image_ids[d]]
        bb = BB[d, :].astype(float) #for VOC
        ovmax = -np.inf
        BBGT = R['bbox'].astype(float)

        if BBGT.size > 0:
            # compute overlaps
            # intersection
            ixmin = np.maximum(BBGT[:, 0], bb[0])
            iymin = np.maximum(BBGT[:, 1], bb[1])
            ixmax = np.minimum(BBGT[:, 2], bb[2])
            iymax = np.minimum(BBGT[:, 3], bb[3])
            iw = np.maximum(ixmax - ixmin + 1., 0.)
            ih = np.maximum(iymax - iymin + 1., 0.)
            inters = iw * ih

            # union
            uni = ((bb[2] - bb[0] + 1.) * (bb[3] - bb[1] + 1.) + 
                   (BBGT[:, 2] - BBGT[:, 0] + 1.) * 
                   (BBGT[:, 3] - BBGT[:, 1] + 1.) - inters) 

            overlaps = inters / uni
            ovmax = np.max(overlaps)
            jmax = np.argmax(overlaps)

        if ovmax > ovthresh:
            if not R['difficult'][jmax]:
                if not R['det'][jmax]:
                    tp[d] = 1.
                    R['det'][jmax] = 1
                else:
                    fp[d] = 1.
        else:
            fp[d] = 1.

    # compute precision recall
    fp = np.cumsum(fp)
    tp = np.cumsum(tp)
    rec = tp / float(npos)
    # avoid divide by zero in case the first detection matches a difficult
    # ground truth
    prec = tp / np.maximum(tp + fp, np.finfo(np.float64).eps)
    ap = voc_ap(rec, prec, use_07_metric)

    return rec, prec, ap
